### PR TITLE
Render sameInstance properites inline

### PR DESCRIPTION
**Pull request by aurislt7**

### CODE REVIEW

File: ui/components/src/ui-components/table/component/TableCell.tsx
Comment (grigasp) [inline ui/components/src/ui-components/table/component/TableCell.tsx:40]: Why not do that through style prop?

File: ui/components/src/ui-components/table/TableDataProvider.ts
Comment (grigasp) [reply ui/components/src/ui-components/table/TableDataProvider.ts:84]: Also, please set this to `@alpha` for now

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:49]: The type should be `RelationshipMeaning`, right?

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:382]: This is a breaking change. I'd say let's just leave the public member without ability to set it through constructor.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:272]: Do you really need a copy here? If yes, you should use `clone` function for that.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:266]: It seems the function is used for two different purposes:
- replace same instance nested content fields with their nested fields (for `createColumns`)
- find same instance nested content fields (for `createRow` call)

Common pieces should be reused, but we should not be doing unnecessary work. E.g. we don't need to be creating `foundFields` map when calling it from `createColumns` and we don't need to do replacements when calling from `createRow`.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:268]: As far as I can tell this won't handle the case when there's a "same instance" nested content field `A` nested inside another similar field `B` - we're going to replace `B` with `A` and increase the index, which means we're not going to handle `A`.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:293]: Instead of `! as NestedContentValue[]` we should use `assert(Value.isNestedContent(value))`

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:290]: `fieldNames` instead of `fields`?

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:301]: I don't understand what's going on here. Can you explain?

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:370]: I'd set these two attributes only if `mergedCellsCount !== undefined`. This would remove those empty changes from our snapshots.

**Review by grigasp — CHANGES_REQUESTED** [review]
Please also add a couple of full-stack tests (one for expanded same-instance case and one for merged cells case)

File: ui/components/src/ui-components/table/component/TableCell.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/TableCell.tsx]: Moved to css

File: ui/components/src/ui-components/table/component/TableCell.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/TableCell.tsx]: Moved to style props

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts:267]: Updated

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Changed

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Changed

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Added comments in the code

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Updated

File: ui/components/src/ui-components/table/component/TableCell.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/TableCell.tsx]: Moved to style prop

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Fields.ts:50]: Changed

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (aurislt7) [reply presentation/common/src/presentation-common/content/Fields.ts]: Updated

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: I just needed to reset parentship. Updated

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts:266]: Updated to have two different functions for each task

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:10]: should not import from barrel file of the same package

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [reply presentation/common/src/presentation-common/content/Fields.ts:50]: still a `string` in `NestedContentFieldJSON`

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:431]: should be unnecessary

File: presentation/common/src/presentation-common/content/Fields.ts
Comment (grigasp) [inline presentation/common/src/presentation-common/content/Fields.ts:458]: should be unnecessary

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Why do you need to reset parentship? That mutates the descriptor which we don't want.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: This is very hard to read.. I'd rewrite like this:
```ts
  return {
    key,
    cells: descriptor.fields.map((field: Field): CellItem => {
      let itemProps: Partial<CellItem> = {};
      const mergedCellsCount = mergedCellsCounts[field.name];
      if (mergedCellsCount) {
        itemProps.mergedCellsCount = mergedCellsCount;
        itemProps.alignment = HorizontalAlignment.Center;

        const nestedField = sameInstanceFieldsMap[field.name];
        nestedField.name = field.name;
        field = nestedField;
      }
      return {
        key: field.name,
        record: ContentBuilder.createPropertyRecord({ field }, item).record,
        ...itemProps,
      };
    }),
  };
```

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Yes, updated it to return a new list

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Sure, updated it to return a new ValuesDictionary

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Yeh. Looks a lot better, thanks!

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: We don't anymore, deleted

File: ui/components/src/ui-components/table/TableDataProvider.ts
Comment (aurislt7) [reply ui/components/src/ui-components/table/TableDataProvider.ts]: Updated

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Updated

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Not needed anymore, deleted.

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Updated so that column width would be read from the gridRef, instead of state.columns.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Now it returns a new list, but the fields themselves still get mutated

File: ui/components/src/ui-components/table/component/Table.scss
Comment (grigasp) [inline ui/components/src/ui-components/table/component/Table.scss:134]: You inserted this new style between a comment and the style that it's commenting. Please move the new style up/down and add a comment explaining why the override is needed.

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (grigasp) [inline ui/components/src/ui-components/table/component/Table.tsx:392]: I'd move this down under data provider checks. Also, please add a comment explaining why `forceUpdate` is needed

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (grigasp) [inline ui/components/src/ui-components/table/component/Table.tsx:1050]: I'd change this to 
```ts
// istanbul ignore else
if (this._gridRef.current && (this._gridRef.current as any).getDataGridDOMNode)
  gridColumns = (this._gridRef.current as any).getDataGridDOMNode().querySelectorAll(".react-grid-HeaderCell");
```

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (grigasp) [inline ui/components/src/ui-components/table/component/Table.tsx:1111]: this should be wrapped with curly braces

**Review by grigasp — CHANGES_REQUESTED** [review]
Several more things:
- We need to wait until `imodel02` is merged with `master` and then merge this branch with it. There will be conflicts related to `Field.relationshipMeaning`.
- Please review the whole PR for style and code readability. Especially changes in `presentation\components\src\presentation-components\table\DataProvider.ts`.

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts]: Updated new list to create field clones

File: ui/components/src/ui-components/table/component/Table.scss
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.scss:134]: Fixed

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Fixed

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Changed it, looks better. Thanks

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx:1116]: Added curly braces

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (grigasp) [inline presentation/components/src/presentation-components/table/DataProvider.ts:393]: We're doing this for every row, although the result is always the same. Could we do it once and share the result between all rows?

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (grigasp) [inline presentation/components/src/test/table/DataProvider.test.ts:20]: lint

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (grigasp) [inline presentation/components/src/test/table/DataProvider.test.ts:444]: typo: `sameInstace`

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (grigasp) [inline presentation/components/src/test/table/DataProvider.test.ts:511]: typo: `sameInstace`

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (grigasp) [inline presentation/components/src/test/table/DataProvider.test.ts:573]: please add a similar test for this case:
```ts
      const values = {
        [nestedField.name]: [],
      };
```

File: presentation/components/src/test/table/DataProvider.test.snap
Comment (grigasp) [inline presentation/components/src/test/table/DataProvider.test.snap:41]: I don't see a test for this snapshot

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts:265]: Fixed it

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts:305]: Looks better. Changed it, thanks

File: presentation/components/src/presentation-components/table/DataProvider.ts
Comment (aurislt7) [reply presentation/components/src/presentation-components/table/DataProvider.ts:393]: Sure, updated the code so that now we would get `sameInstanceFieldsMap` and `updatedFields` in createRows function and pass them for every row from there.

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (aurislt7) [reply presentation/components/src/test/table/DataProvider.test.ts:20]: fixed

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (aurislt7) [reply presentation/components/src/test/table/DataProvider.test.ts:444]: fixed

File: presentation/components/src/test/table/DataProvider.test.ts
Comment (aurislt7) [reply presentation/components/src/test/table/DataProvider.test.ts:511]: fixed

File: presentation/components/src/test/table/DataProvider.test.snap
Comment (aurislt7) [reply presentation/components/src/test/table/DataProvider.test.snap:41]: Deleted

**Review by grigasp — CHANGES_REQUESTED** [review]
A couple of last issues:
- need to merge with master
- need to run `extract-api`
- shouldn't reference stuff with square brackets in change log JSONs - instead, wrap them with <code>`</code>

**Review by grigasp — APPROVED** [review]

**Comment by bsteinbk** [comment]
@aurislt7 - would it be possible for you to add an example table, maybe from your test code, that demonstrates the table and this new functionality in the Component Examples stage in ui-testapp? See other examples at https://github.com/imodeljs/imodeljs/commit/cc1864e3d643015ffa71a697030c94db14ea4926

**Review by bsteinbk — APPROVED** [review]

File: ui/components/src/ui-components/table/component/Table.tsx
Comment (saskliutas) [inline ui/components/src/ui-components/table/component/Table.tsx:1162]: Have you checked `classnames` package? Maybe it has function to concatenate class names that you could use?

Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx:1162]: Yep, checked it, however function provided by `classnames` package stacked classnames on top of each other even though some of the classnames were already set. I could use it, but then the final classname comes out clustered with repeated strings.

Comment (saskliutas) [reply ui/components/src/ui-components/table/component/Table.tsx:1162]: It looks like this should work https://github.com/JedWatson/classnames#alternate-dedupe-version

Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Yeep. Missed it, thanks!

Comment (saskliutas) [reply ui/components/src/ui-components/table/component/Table.tsx]: I guess it would be better to import both versions of `classnames` function (`"classnames"` and `"classnames/dedupe"`) and use dedupe version only where you used `mergeClassNames` and leave all other places to use `classnames` without dedupe as it was before.

Comment (aurislt7) [reply ui/components/src/ui-components/table/component/Table.tsx]: Makes sense. Updated

**Pull request merged**
